### ISSUE TITLE
soundcloud integration

### ISSUE THREAD

**Issue by muzzol** [tootsuite/mastodon#921]
would like to see it integrated with soundcloud so you can see the full frame with song instead of just a link.

there's some discussion here: https://github.com/tootsuite/mastodon/issues/309
but I'm not sure how is related to my feature request.

**Comment by bitfehler** [comment]
I was looking into supporting oEmbed in a web service. I would prefer oEmbed over e.g. OpenGraph due to the separation of the oEmbed endpoint from the actual endpoints that might be linked to, as it has the potential to ease the load on the actual endpoints. Previews on Mastodon are a first class target for this.

I then found out that Mastodon in fact [does not accept type `rich`](https://github.com/mastodon/mastodon/blob/6cf57c6/app/services/fetch_link_card_service.rb#L138-L140), with a comment stating

> Most providers rely on <script> tags, which is a no-no

I found that a little odd, but do have some sympathy for that sentiment. OEmbed does provide the `link` type, where the styling is left to the client (e.g. Mastodon in this case), which isn't great, but certainly has some benefits for users/clients. I was curious how Mastodon renders these, though. Not having seen a single provider provide this type so far, I figured Mastodon itself would probably be the right place to look. However:

```
$ curl -Ss 'https://mastodon.social/api/oembed?format=json&url=https%3A%2F%2Fmastodon.social%2F%40Gargron%2F109366815553433722' | jq .
{
  "type": "rich",
  "version": "1.0",
  "author_name": "Eugen Rochko",
  "author_url": "https://mastodon.social/@Gargron",
  "provider_name": "mastodon.social",
  "provider_url": "https://mastodon.social/",
  "cache_age": 86400,
  "html": "<iframe src=\"https://mastodon.social/@Gargron/109366815553433722/embed\" class=\"mastodon-embed\" style=\"max-width: 100%; border: 0\" width=\"400\" allowfullscreen=\"allowfullscreen\"></iframe><script src=\"https://static-cdn.mastodon.social/embed.js\" async=\"async\"></script>",
  "width": 400,
  "height": null
}
```

So not only does Mastodon itself use the `rich` type, it even uses a `<script>` tag? I'll refrain from snarky comments, as it's easy to see how this goes down in a project as huge a Mastodon. But I'd say it's not ideal. In any case, it seems Mastodon does not support its own oEmbed endpoint?

I then found this issue, and I honestly don't want to just rant here, I am genuinely interested in a solution, but I am also somewhat confused by this. To outright block type `rich` seems overly heavy handed, as it is by far the most widely used type out there. The providers want some control over how their previews get displayed. I could understand the stance to deny this _if_ Mastodon itself were to lead by example.

So, my question would be: what would it take to get this feature accepted? The comment in the code does not really instill much confidence that it ever would be, but given this issue here, it seems like there is a chance?

The `video` type also supplies custom HTML, but instead of being outright rejected, there is some sanitization happening. Would that be an option? Or, given how Mastodon's own oEmbed endpoint is implemented, maybe it's fine to just allow it and let user-agents handle this? For example, the above mentioned SoundCloud oEmbed is blocked by default by at least Privacy Badger, and probably other such extensions, also.

PS: Thanks a lot for Mastodon :wink:

**Comment by afontenot** [comment]
A few notes here.

1. On the "script" issue. Script is OK if it's inside a page fetched via an iframe src, because a remote resource won't be able to access the page because of browser protections. Videos often use script inside of an iframe like this.

2. For "rich" specifically, I *think* the usual approach here is to have another domain that hosts the HTML content and the home domain embeds it via an iframe. That solves a bunch of security concerns but it might not be possible to implement in Mastodon (which needs to be usable by casual sysadmins on a single domain).
\
Sanitation could *potentially* solve some of these problems but it's inherently risky, and there would be no way for Mastodon to know whether it was working on not. A whole bunch of rich embeds would probably just break because of broken JavaScript. Sanitation has the potential to break video embeds as well, but we do it anyway in the hope that most video sources are smart and don't use script tags.

3. For someone creating a project specifically aimed at Mastodon, you can actually do a lot by simply pretending that your content is video. As long as you can pass the sanitizer it's no problem. As an example, I have [a site that lets you embed pastes from Github and Pastebin with working syntax highlighting](https://paste.social/) which uses this trick.

**Comment by bitfehler** [comment]
Thank you for getting back!

>   1. On the "script" issue. Script is OK if it's inside a page fetched via an iframe src, because a remote resource won't be able to access the page because of browser protections. Videos often use script inside of an iframe like this.

In Mastodon's oembed HTML snippet, the script tag is outside of the iframe, though? But it also doesn't matter, it just struck me as odd, but I'm more interested in oembedding of other content rather than how Mastodon provides oembed.

>   2. For "rich" specifically, I _think_ the usual approach here is to have another domain that hosts the HTML content and the home domain embeds it via an iframe. That solves a bunch of security concerns but it might not be possible to implement in Mastodon (which needs to be usable by casual sysadmins on a single domain).

This is of course very important, though I think there might be a better approach, see below.

>    Sanitation could _potentially_ solve some of these problems but it's inherently risky, and there would be no way for Mastodon to know whether it was working on not. A whole bunch of rich embeds would probably just break because of broken JavaScript. Sanitation has the potential to break video embeds as well, but we do it anyway in the hope that most video sources are smart and don't use script tags.
> 
>  3. For someone creating a project specifically aimed at Mastodon, you can actually do a lot by simply pretending that your content is video. As long as you can pass the sanitizer it's no problem. As an example, I have [a site that lets you embed pastes from Github and Pastebin with working syntax highlighting](https://paste.social/) which uses this trick.

Just to give my opinion on this: I find it a bit confusing. If I understand correctly, there are two issues: a security concern (random content displayed in a trusted domain), and a usability issue (a sanitized snippet potentially not working). However, your suggested work-around clearly shows that there is no security benefit from not allowing `rich` at all, because (e.g. as an attacker) I can do the same thing via `video`. That leaves usability. And not displaying well-behaved oEmbed content at all because other content _might_ break already seems a little odd to me. But encouraging work-arounds like that is - I would honestly say - clearly not doing a favor to the usability of the entire oEmbed eco-system.

Now, again, I'd love to see this sorted out, so let me make some suggestions:

- Currently, [oEmbed is preferred over OpenGraph](https://github.com/mastodon/mastodon/blob/48e136605a30fa7ee71a656b599d91adf47b17fc/app/services/fetch_link_card_service.rb#L42). I don't know why that is, but how about this gets switched around? This would reduce the surface of the problem to only the providers that provide oembed but not opengraph, which I think are a _lot_ less already. 
- With that in place, maybe it might already make sense to just enable `rich` and use the same sanitization as `video`?
- I admit that my web security heyday is some time ago, but: wouldn't the "proper" solution be to display the remote content in a [sandboxed iframe](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe#attr-sandbox)? That will essentially treat the contents of the iframe as if it came from a different origin. You could even pass the HTML in the [srcdoc attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe#attr-srcdoc), so it never has to be served as actual content from the origin domain.

**Comment by afontenot** [comment]
> However, your suggested work-around clearly shows that there is no security benefit from not allowing `rich` at all, because (e.g. as an attacker) I can do the same thing via `video`. That leaves usability. And not displaying well-behaved oEmbed content at all because other content _might_ break already seems a little odd to me.

Right, that's correct. I think it's just a case of trying to match user expectations. Following the sanitation approach for video content will almost always work - the user will be able to display the video. The same approach for rich content will often (usually?) not work; it will show something but it will be broken. Trying to change sanitation to get most rich content working would be risky or might not work at all. The reason embeds of Mastodon content are "allowed" to use `<script>` is that they're embedded by users who trust Mastodon on their own sites, or they're wrapped in a protective iframe by other sites.

> Currently, [oEmbed is preferred over OpenGraph](https://github.com/mastodon/mastodon/blob/48e136605a30fa7ee71a656b599d91adf47b17fc/app/services/fetch_link_card_service.rb#L42). I don't know why that is, but how about this gets switched around?

Oembed is a lot more powerful than OG. The latter usually provides only a static representation of the content. For example YouTube provides both, but only Oembed will give you a video playable without leaving the Mastodon interface.

I like your idea of using a sandboxed iframe here. Actually, it would drastically lower the attack surface even on video oembeds. As it stands, one screwup in the (very loose!) sanitizer configuration (or a vulnerability in the library) and it's game over for Mastodon. If a sandboxed iframe can reliably get the browser to treat all this user-content as other-origin, that's a good move IMO.

**Comment by Floppy** [comment]
Seconding this please! I've just done a demo of a Manyfold 3d view showing in the Mastodon timeline, but currently I need to lie and say it's a `video`, not `rich` (which I'm not releasing, this was just an experiment). There's no difference at all in security between rich and video, both use the HTML sent over by the client, so it could work in exactly the same way that video sandboxing presumably works at the moment? 

https://3dp.chat/@manyfold/114160446378845640

I'm happy to do a PR, if it's likely to be accepted. 🤞🏻

**Comment by winie** [comment]
Perhaps it would be a good compromise if OEmbed ‘rich’ type were initially interpreted as ‘link’ type and the html field were ignored.

This would allow WordPress links to function initially. At least an image and further information would be displayed.

What are your opinions?
